### ISSUE TITLE
Log level with extra context information where the context affects the log level behaviour

### ISSUE THREAD

**Issue by symbolix** [getlogbook/logbook#227]
Hi,

Thank you very much for logbook. I am curious about a specific user case and would like to know if that is possible with logbook.

I would like to set-up a custom log level that besides displaying some contextual information also is controlled by that very same contextual information. Let me sum this up in an example.

Here is a function that is being logger:

```python
def test_condition_info():
    my_log.event('Limit event has occured.', current_size=11,size_limit=9)
```

and the output is:

`[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (EVENT) Limit event has occurred. >> {"current_size":11, "size_limit":9}`

and the reason we see this message is not only because the level is set to  logging.EVENT but also because of a pre-defined logic `if current_size > size_limit: register the message`.

In a similar fashion a call like this one:

```python
my_log.event('Limit event has occurred.', current_size=5, size_limit=9)
```

will result with the message being NOT displayed even the log level is set to EVENT, since the `if current_size > size_limit: register the message` is not met.

I know that this can be handled within the function with an `if block`, however, my case is slightly more complex and this is what I need to achieve.

Is this kind of a customisation possible with logbook? Thanks.

**Comment by vmalloc** [comment]
The current implementation assumes the `level` is fixed and static. I kind of lost you on what you really want to achieve, but I doubt we will make any complex logical changes to how records are filtered and/or heavy-initialized...

**Comment by symbolix** [comment]
Hi,

Thanks for responding to my issue. I am not relying on changing levels etc. In fact, probably what I want to achieve should be possible with **logbook**, however I am not sure how exactly.

Is it possible to pass contextual information with the log messages? As far as I understand it is.

`my_log.debug('This is a debug message.', current_size=5, size_limit=9)`

and the output is:

`[2016-10-08] <__main__> {myApp:test_condition_info:93} (DEBUG) This is a debug message. >> {"current_size":5, "size_limit":9}`

Is the case above possible with **logbook**?

**Comment by vmalloc** [comment]
I added an example to a new cookbook section - let me know if this answers your question: http://logbook.readthedocs.io/en/latest/cookbook.html

**Comment by symbolix** [comment]
Thanks for adding the example. It might be close to what I am trying to achieve. I have included a test case that describes what I am trying to achieve:

I would like to extend the existing logging.LEVEL mechanics so that I have the option of switching between different logging levels such as DEBUG, INFO, ERROR etc. (which is a standard feature) but also define a depth for each of the levels.

For example, let's assume that the logging level is set to logging.DEBUG. All of the logbook.DEBUG() calls will be visible as expected.

```python
logbook.debug('Limit event has occurred.')
```

So I get:

```
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) Limit event has occurred.
```

What I am after, is passing an extra depth level to the `logbook.debug()` call so that I can control how much of detail is printed in the DEBUG message, not entirely enabling or disabling the DEBUG level but controlling how much information the debug message will carry. So in all cases we see the debug message but in some instances, it is less detailed and on some occasions more information is included.

For example:

```python
logbook.debug('Limit event has occurred.', verbosity=1)
logbook.debug('The following user has caused the limit event: %s' % (user), verbosity=3)
logbook.debug('The following files have been affected: %s' % [list_of_files], verbosity=7)
```

So when the logging level is set to DEBUG and the global verbosity is set to `GLOBAL_VERBOSITY=1` we will get this:

```
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) Limit event has occurred.
```

And if the global verbosity is set to `GLOBAL_VERBOSITY=4` we will get this:

```
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) Limit event has occurred.
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) The following user has caused the limit event: xpr
```

And if the global verbosity is set to `GLOBAL_VERBOSITY=9` we will get all of the details:

```
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) Limit event has occurred.
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) The following user has caused the limit event: xpr
[2016-10-08 10:07:29,807] <__main__> {myApp:test_condition_info:93} (DEBUG) The following files have been affected: ['inside.ini', 'render.so']
```

I am not sure if this is possible with **logbook** out of the box or do I need to create some customisation. Thank you very much.

**Comment by vmalloc** [comment]
The way I see it you have to write a custom handler to achieve this
behavior. It isn't so hard to do.

**Comment by symbolix** [comment]
Ah, thanks for the encouraging words :)

I was starting to get a bit depressed by this issue. So do I inherit from `logbook.HANDLERS` and create a custom `handler` class?

Thank you very much.

**Comment by vmalloc** [comment]
Yeah. Take a look at FileHandler, and you can inherit from it and override
the format method to suit your needs.

**Comment by symbolix** [comment]
Ok. Thanks. Where would be the best place to insert the conditional logic?

```python
if verbosity < GLOBAL_VERBOSITY:
    display_the_message
else:
    do_not_display_the_message
```

Inside the actual `logbook.debug()` function?

**Comment by vmalloc** [comment]
No, because debug isn't under your control. It should be in the format
method.
